wandb/sdk/launch/environment/gcp_environment.py

```diff
@@ -0,0 +1,229 @@
+"""Implementation of the GCP environment for wandb launch."""
+import logging
+import os
+import re
+
+from wandb.errors import LaunchError
+from wandb.util import get_module
+
+from .abstract import AbstractEnvironment
+
+google = get_module(
+    "google",
+    required="Google Cloud Platform support requires the google package. Please"
+    " install it with `pip install wandb[launch]`.",
+)
+google.cloud.compute_v1 = get_module(
+    "google.cloud.compute_v1",
+    required="Google Cloud Platform support requires the google-cloud-compute package. "
+    "Please install it with `pip install wandb[launch]`.",
+)
+google.auth.credentials = get_module(
+    "google.auth.credentials",
+    required="Google Cloud Platform support requires google-auth. "
+    "Please install it with `pip install wandb[launch]`.",
+)
+google.auth.transport.requests = get_module(
+    "google.auth.transport.requests",
+    required="Google Cloud Platform support requires google-auth. "
+    "Please install it with `pip install wandb[launch]`.",
+)
+google.api_core.exceptions = get_module(
+    "google.api_core.exceptions",
+    required="Google Cloud Platform support requires google-api-core. "
+    "Please install it with `pip install wandb[launch]`.",
+)
+google.cloud.storage = get_module(
+    "google.cloud.storage",
+    required="Google Cloud Platform support requires google-cloud-storage. "
+    "Please install it with `pip install wandb[launch].",
+)
+
+
+_logger = logging.getLogger(__name__)
+
+gcs_uri_re = re.compile(r"gs://([^/]+)/(.+)")
+
+
+class GcpEnvironment(AbstractEnvironment):
+    """GCP Environment.
+
+    Attributes:
+        region: The GCP region.
+    """
+
+    region: str
+    _project: str
+
+    def __init__(self, region: str, verify: bool = True) -> None:
+        """Initialize the GCP environment.
+
+        Args:
+            region: The GCP region.
+            verify: Whether to verify the credentials, region, and project.
+
+        Raises:
+            LaunchError: If verify is True and the environment is not properly
+                configured.
+        """
+        super().__init__()
+        _logger.info(f"Initializing GcpEnvironment in region {region}")
+        self.region: str = region
+        self._project = ""
+        if verify:
+            self.verify()
+
+    @property
+    def project(self) -> str:
+        """Get the name of the gcp project.
+
+        The project name is determined by the credentials, so this method
+        verifies the credentials if they have not already been verified.
+
+        Returns:
+            str: The name of the gcp project.
+
+        Raises:
+            LaunchError: If the launch environment cannot be verified.
+        """
+        if not self._project:
+            raise LaunchError(
+                "This GcpEnvironment has not been verified. Please call verify() "
```

Comment:
Couldn't they also call `get_credentials`?

Reply:
If get_credentials is only to be called by the class in its other methods, we can prepend it with `_` to make that more clear

Reply:
I guess you're right but in either case then _project would be set before. I don't think that this will every be hit honestly, I added the verify=True argument because I figured it might come in handy one day, and maybe with tests. But it will not called with verify=False in our cli/sdk code

Reply:
Sounds good